cirq/ops/projectors.py

```diff
@@ -0,0 +1,75 @@
+from typing import Any, cast, Dict, List, Tuple, TYPE_CHECKING, Union
+
+import numpy as np
+
+from cirq import value
+from cirq.ops import raw_types
+
+if TYPE_CHECKING:
+    import cirq
+
+
+@value.value_equality
+class Projector(raw_types.Gate):
+    """A non-unitary gate that projects onto a single qubit."""
+
+    def __init__(self,
+                 projection_basis: Union[List[List[float]], np.ndarray],
+                 qid_shape: Tuple[int, ...] = (2,)):
+        """
+        Args:
+            projection_basis: a (2**num_qubits, p) matrix that lists the
+                projection vectors.
+            qid_shape: Specifies the dimension of the projection. The default is
+                2 (qubit).
+
+        Raises:
+            ValueError: If the basis vector is empty.
+        """
+        if type(projection_basis) == np.ndarray:
+            projection_array = cast(np.ndarray, projection_basis)
+        else:
+            projection_array = np.array(projection_basis)
+
+        if np.prod(qid_shape) != projection_array.shape[1]:
```

Comment:
I'm not sure what you mean by 'rank' here. Do you mean that we should check that the basis spans a 2D space (i.e. the rank of the matrix is 2) or do you mean that the input array has shape[0] == 2?

In either case, I don't understand why it would be a requirement (not contesting, but rather not understanding).

Leaving unresolved.

Reply:
Sorry I wasn’t all that clear. I was just thinking in terms of error reporting if the user accidentally called the function with  a 1D or 3D array they might get a cryptic failure is all and should we catch that case or not ?

Reply:
Oh, that makes a lot of sense. Done + added a test. Thank you.